librz/bin/dwarf/line.c

```diff
@@ -0,0 +1,774 @@
+ */
+static size_t RzBinDwarfLineHeader_std_opcode_args_count(
+	const RzBinDwarfLineHeader *hdr, ut8 opcode) {
+	if (!opcode || opcode > hdr->opcode_base - 1 || !hdr->std_opcode_lengths) {
+		return 0;
+	}
+	return hdr->std_opcode_lengths[opcode - 1];
+}
+
+static bool RzBinDwarfLineOp_parse_std(
+	RzBuffer *buffer,
+	RzBinDwarfLineOp *op,
+	const RzBinDwarfLineHeader *hdr,
+	enum DW_LNS opcode,
+	bool big_endian) {
+	rz_return_val_if_fail(op && hdr && buffer, false);
+	op->type = RZ_BIN_DWARF_LINE_OP_TYPE_STD;
+	op->opcode = opcode;
+	switch (opcode) {
+	case DW_LNS_advance_pc:
+		ULE128_OR_RET_FALSE(op->args.advance_pc);
+		break;
+	case DW_LNS_advance_line:
+		SLE128_OR_RET_FALSE(op->args.advance_line);
+		break;
+	case DW_LNS_set_file:
+		ULE128_OR_RET_FALSE(op->args.set_file);
+		break;
+	case DW_LNS_set_column:
+		ULE128_OR_RET_FALSE(op->args.set_column);
+		break;
+	case DW_LNS_fixed_advance_pc:
+		U_OR_RET_FALSE(16, op->args.fixed_advance_pc);
+		break;
+	case DW_LNS_set_isa:
+		ULE128_OR_RET_FALSE(op->args.set_isa);
+		break;
+
+	// known opcodes that take no args
+	case DW_LNS_copy:
+	case DW_LNS_negate_stmt:
+	case DW_LNS_set_basic_block:
+	case DW_LNS_const_add_pc:
+	case DW_LNS_set_prologue_end:
+	case DW_LNS_set_epilogue_begin:
+		break;
+
+	// unknown operands, skip the number of args given in the header.
+	default: {
+		size_t args_count = RzBinDwarfLineHeader_std_opcode_args_count(hdr, opcode);
+		for (size_t i = 0; i < args_count; i++) {
+			ULE128_OR_GOTO(op->args.advance_pc, ok);
+		}
+	}
+	}
+ok:
+	return true;
+}
+
+RZ_IPI void RzBinDwarfSMRegisters_reset(
+	const RzBinDwarfLineHeader *hdr,
+	RzBinDwarfSMRegisters *regs) {
+	rz_return_if_fail(hdr && regs);
+	regs->address = 0;
+	regs->file = 1;
+	regs->line = 1;
+	regs->column = 0;
+	regs->is_stmt = hdr->default_is_stmt;
+	regs->basic_block = DWARF_FALSE;
+	regs->end_sequence = DWARF_FALSE;
+	regs->prologue_end = DWARF_FALSE;
+	regs->epilogue_begin = DWARF_FALSE;
+	regs->isa = 0;
+}
+
+static void store_line_sample(
+	RzBinSourceLineInfoBuilder *bob,
+	const RzBinDwarfLineHeader *hdr,
+	RzBinDwarfSMRegisters *regs,
+	RZ_NULLABLE RzBinDwarfDebugInfo *info,
+	RZ_NULLABLE RzBinDwarfLineFileCache *fnc) {
+	const char *file = NULL;
+	if (regs->file) {
+		file = get_full_file_path(info, hdr, fnc, regs->file - 1);
+	}
+	rz_bin_source_line_info_builder_push_sample(bob, regs->address, (ut32)regs->line, (ut32)regs->column, file);
+}
+
+/**
+ * \brief Execute a single line op on regs and optionally store the resulting line info in bob
+ * \param fnc if not null, filenames will be resolved to their full paths using this cache.
+ */
+RZ_IPI bool RzBinDwarfLineOp_run(
+	const RzBinDwarfLineHeader *hdr,
+	RzBinDwarfSMRegisters *regs,
+	RzBinDwarfLineOp *op,
+	RZ_NULLABLE RzBinSourceLineInfoBuilder *bob,
+	RZ_NULLABLE RzBinDwarfDebugInfo *info,
+	RZ_NULLABLE RzBinDwarfLineFileCache *fnc) {
```

Comment:
Those are quite a lot parameters. Consider putting as many of them in a struct and pass this one. Also because other functions below use the same set of parameters as well.